http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/sam/SamServiceInterp.scala

```diff
@@ -135,4 +164,126 @@ class SamServiceInterp[F[_]](apiClientProvider: SamApiClientProvider[F],
           F.raiseError(LeoInternalServerError("Internal server error retrieving Leo credentials", Some(ctx.traceId)))
       }
     } yield token
+
+  private def isAuthorized(bearerToken: String, samResourceId: SamResourceId, action: String)(implicit
+    ev: Ask[F, AppContext]
+  ): F[Boolean] =
+    for {
+      ctx <- ev.ask
+      resourcesApi <- apiClientProvider.resourcesApi(bearerToken)
+      isAuthorized <- SamRetry
+        .retry(resourcesApi.resourcePermissionV2(samResourceId.resourceType.asString, samResourceId.resourceId, action),
+               "resourcePermissionV2"
+        )
+        .adaptError { case e: ApiException =>
+          SamException.create("Error checking resource permission in Sam", e, ctx.traceId)
+        }
+
+    } yield isAuthorized
+
+  override def checkAuthorized(bearerToken: String, samResourceId: SamResourceId, action: String)(implicit
+    ev: Ask[F, AppContext]
+  ): F[Unit] = for {
+    ctx <- ev.ask
+    isAuthorized <- isAuthorized(bearerToken, samResourceId, action)
+    userEmail <- getUserEmail(bearerToken)
+    _ <- F.raiseWhen(!isAuthorized)(
+      SamException.create(
+        s"User $userEmail is not authorized to perform action $action on ${samResourceId.resourceType} ${samResourceId.resourceId}",
+        StatusCodes.Forbidden.intValue,
+        ctx.traceId
+      )
+    )
+    _ <- logger.info(ctx.loggingCtx)(
+      s"User $userEmail is authorized to $action ${samResourceId.resourceType} ${samResourceId.resourceId}"
+    )
+  } yield ()
+
+  override def listResources(bearerToken: String, samResourceType: SamResourceType)(implicit
+    ev: Ask[F, AppContext]
+  ): F[List[String]] = for {
+    ctx <- ev.ask
+    resourcesApi <- apiClientProvider.resourcesApi(bearerToken)
+    resources <- SamRetry
+      .retry(resourcesApi.listResourcesAndPoliciesV2(samResourceType.asString), "listResourcesAndPoliciesV2")
+      .adaptError { case e: ApiException =>
+        SamException.create("Error listing resources from Sam", e, ctx.traceId)
+      }
+  } yield resources.asScala.toList.map(_.getResourceId)
```

Comment:
"has access to" might be a bit limiting. I don't know that there is a use case for listing just the resource IDs here - there might be a specific access right that we want to check, like `read`. I don't think we necessarily need to change that in this PR, since this method isn't actually used yet. Just a headsup that this might need to change for Leo's use case

Reply:
Yeah that's a good point - we might want to additionally check actions here. I agree this can be changed in a subsequent PR.

Reply:
So y ou are saying adding a `scope` to this function that could be a role or an action? E.g. list all resources the user has read access to? That might be tricky since different resources can have completely different actions no?

Reply:
I think ideally the current use cases would be covered by adding a list of actions. We don't want to check roles if we can avoid it (we might have to if we have to use parent role as a standin for child action). This method is only called in the context of a single resource type so that will scope us to only the actions for that type.

Reply:
+1 -- I'm imagining adding a new parameter `actions: List[SamAction]` to this method, which would further constrain the results.